### ISSUE TITLE
Capacity registration flow 3Bot tfchain prototype

### ISSUE THREAD

**Issue by LeeSmet** [threefoldtech/home#52]
Allow a user to pay and use capacity on the grid by sending money and the capacity order to a 3Bot. Testnet prototype only for now.

**Comment by LeeSmet** [comment]
![img_20181126_101030](https://user-images.githubusercontent.com/22023708/49004256-25eb4780-f164-11e8-8a76-30d95b370895.jpg)

**Comment by zaibon** [comment]
Actually I don't think the flow is 100% correct in this schema.
How we were planning to do things is:
- user have it's own 3bot. User wallet is managed by the 3bot
- using 3bot interface, user wants to reserve capacity
- 3bot sends transaction to node robot of the selected nodes
- node robot tries to deploy the capacity
  -  if success, service information is sent back to the 3bot
  - if failure, refund to the wallet address who sent the reservation transaction

**Comment by GlenDC** [comment]
Sure, but from the low-level blockchain POV that is still close enough. Wether a 3bot sends that money+order or it is a user, is no difference, as the 3Bot can been seen as a user.

> refund to the wallet address who sent the reservation transaction

This refund will not be 100% though. The transaction fee will be subtracted twice,
given the node-3Bot cannot pay for these fees. Otherwise the node-3Bot would be paying those fees out of its pocket, which doesn't seem right for multiple reasons.

Still the refund will be close enough to the full amount. For the real version we probably do want to prevent this if possible (by allowing somehow these pre-checks up-front off-chain).

**Comment by GlenDC** [comment]
Besides the high level difference that there is a `user <-> personal-3Bot` intersection, does it look like this could work for a first `testnet` prototype @zaibon?

**Comment by zaibon** [comment]
> Wether a 3bot sends that money+order or it is a user, is no difference, as the 3Bot can been seen as a user.

Very true, just wanted to highlight that the 3bot is the one going to manager the wallet and create the transaction.

> Besides the high level difference that there is a user <-> personal-3Bot intersection, does it look like this could work for a first testnet prototype @zaibon?

Yep, I think it can work

Regarding the refund flow, we should indeed try to get as pre-check as possible to avoid to have to refund at all.

**Comment by GlenDC** [comment]
Further, @zaibon, how do you see the amounts in terms of the units?
* What unit(s) is/are to be used for HDD/SSD: MB, GB, TB?
* What unit(s) is/are to be used for CPU?
* What unit(s) is/are to be used for MEMORY: GB?
* What unit(s) is/are to be used for Network Bandwidth? Or how do you see this unit?

What to do with the duration? You want to express it in terms of a duration (hours, weeks or months), or do you want to simply have it as a unix epoch time stamp, until when it has to remain reserved. In which case it would be 8 bytes for the duration. Or 3 bytes if we apply [the same trick as I apply for our 3Bot expiration time stamps](https://github.com/threefoldfoundation/tfchain/blob/master/doc/binary_encoding.md#compact-timestamp).

For the prototype all have to be fit within 83 bytes (we could increase that upper limit, but given it is a temporary solution I would like to keep it as small as possible. On top of that, the global goal is to keep the footprint of our chain as small as possible, so trying to keep it within the 83 bytes (or even smaller) might help with that goal.

I think it is manageable, if for example you have a first byte which can contain up to 8 flags (1 flag per bit), to indicate what units will be requested. To than have the amounts per requested be defined in binary encoded bytes in order of defined flags. Meaning you could have 79 bytes for all your amounts (83 - 1 (flag byte) - 3 (compact timestamp)), and thus giving you an average of `14.8` bytes per unit (if all 5 units would be defined in one request), which I think should be enough, given that a `uint32` would only require 5 bytes.

To recap it would mean that even if you use a full timestamp (8 bytes) as "duration/deadline", and define all (5) units, which the amount of each unit being close to the uint32 limit, you would still only use 34 bytes of the arbitrary data, giving you still 49 (83 - 1 - 8 - 5*5) that you can use for other purposes should I missed something in the required info of such an order. In theory that could mean you can even add the farmer ID (8 or 32 bytes), but I think if you have one 3Bot per farm, than that might not be needed, as the address to which you pay would already identify the desired farm. Even if not, you clearly have the room for purposes such as that, as you have 49 bytes over to play with, according to my earlier calculation.

**Comment by GlenDC** [comment]
> Regarding the refund flow, we should indeed try to get as pre-check as possible to avoid to have to refund at all.

Sure, and given the order interaction is automated between 3Bots, I don't see why not possible. But in worst case, should the check give a false positive, the user will lose 0.2 TFT because of it. That is for now the consequence.

If your 3Bot-pre-check is done properly that should however be a rare (if even possible) situation, so I wouldn't worry to much about it, especially as for the initial launch in January we will all keep it testnet-only, so people do not lose money (at least that is how I understood it, @robvanmieghem or @despiegk might need to confirm).

**Comment by zaibon** [comment]
See wiki documentation about resource units: https://threefoldfoundation.github.io/info_grid/#/concepts/resource_units

Regrading reservation time I don't know what @despiegk wants to use there.

**Comment by zaibon** [comment]
>  For sure it's more than just about cores

No its not, or at least I've never seen any other definition.

**Comment by GlenDC** [comment]
Ok, all good for me. In that case I think you have all you need than. The data format will be defined somewhere in 3Bot I suppose. Once we're ready for that, you can let me know and I can help by either giving advice on what to do, or I could add some code for it myself, whatever is easiest for you all.

**Comment by GlenDC** [comment]
Tasks to do as to allow flow in tfchain:
- [ ] Review and Merge RIvine+TFchain PRs which add optional typing of arbitrary data (@robvanmieghem, @LeeSmet) (https://github.com/threefoldtech/rivine/pull/473 and https://github.com/threefoldfoundation/tfchain/pull/214);
- [ ] [Support optional typing of arbitrary data in JumpScale light client](https://github.com/threefoldtech/jumpscale_lib/issues/256) (JumpScale, @glendc);

Optimisations to do:
-  [ ] [support min-block-height filter when looking up transactions for unlockhash](https://github.com/threefoldtech/rivine/issues/475) (Rivine, @GlenDC));

**Comment by GlenDC** [comment]
> Define initial capacity order python utility class to easily encode/decode such an order as/from arbitrary data

This is done and implemented in 
https://github.com/threefoldtech/jumpscale_lib/commit/2b5838711896b2ea56672944694c522ae45b2a5a.

Probably you do want to move it into the digitalme library, but you have at least now some code to start from. Implemented as discussed. Here is an example snippet:

```python
from JumpscaleLib.clients.blockchain.tfchain.capacity.order import CapacityOrder

order = CapacityOrder()
# resource units, one or multiple can be defined
order.cru = 1 # expressed in cores
order.mru = 2 # expressed in GB
order.hru = 1000 # expressed in GB
order.sru = 256 # expressed in GB
order.nru = 2 # expressed in GB in/out
# the duration is required as well, expressed in cycles of 30-days
order.duration = 6 # 6*30 days
# optionally you can attach a description,
# will be registered on the chain together with the order,
# maximum 32 bytes
order.description = 'shop A web servers'

# you can print it in the (JS) shell, simply by entering it
order
# will result in:
# Capacity order for 180 days:
#        - CRU: 1 cores
#        - MRU: 2 GB
#        - HRU: 1000 GB
#        - SRU: 256 GB
#        - NRU: 2 GB in/out
#        - description: bytearray(b'shop A web servers')

# encoding to binary can be done using the `binary` method:
order.binary()
# results in:
# bytearray(b'\xf8\x01\x02\x00\xe8\x03\x00\x00\x00\x01\x00\x00\x02\x00\x06\x00\x12shop A web servers')
# which means that such a complete order fits with 35 bytes easily
# in the 83 bytes limit of the arbitrary data property of a tfchain money transaction

# a capacity order can be easily decoded using the `classmethod` from_binary:
assert CapacityOrder.from_binary(bytearray(b'\xf8\x01\x02\x00\xe8\x03\x00\x00\x00\x01\x00\x00\x02\x00\x06\x00\x12shop A web servers')) == order
# and yes, CapacityOrders can be compared, and do so dict-wise property by property
```

I think the result is pretty easy to use. Feel free to modify as you please when moving that python module to the correct library (e.g. digitalme), but it should allow you to quickly encode/decode capacity orders in the initial 3Bot capacity order prototype.

For example you will probably send such a capacity order as follows<sup>(1)</sup>:

```python
mywallet.send_money(amount=price, recipient=node_3bot_address, data=order.binary())
```

---

> <sup>(1)</sup> Once https://github.com/threefoldtech/jumpscale_lib/issues/256 is resolved you probably also need to define the data_type, such that it becomes actually:
> ```python
> mywallet.send_money(amount=price, recipient=node_3bot_address, data=order.binary(), data_type=128)
> ```
> But that design detail is yet to be confirmed by an actual implementation.

**Comment by GlenDC** [comment]
TFchain and JumpScale (tfchain libs) now fully support the prototype as described earlier in this story. Closing this story, given it is considered done. The goals were to figure out what was still to be done on the tfchain side to allow for a capacity reservation prototype, as well as implement all that was missing.
